routing.py

```diff
@@ -0,0 +1,272 @@
+    """
+    Check if this repo has versions
+    """
+
+    if os.path.isfile(version_filepath):
+        with open(version_filepath) as version_file:
+            versions = list(filter(None, version_file.read().splitlines()))
+
+        if len(versions):
+            return versions
+
+
+def get_file(request_path):
+    """
+    Given a request_path, return the path where the file should be
+    """
+
+    if request_path.endswith('/'):
+        return request_path + '/index.html'
+    else:
+        return request_path + '.html'
+
+
+def is_version(version):
+    """
+    Check if a string is one of the available versions
+    """
+
+    versions = get_versions()
+
+    return versions and version in versions
+
+
+def split_path(request_path, languages, versions):
+    """
+    Given a URL path of the form:
+      (/{version})(/language)/remaining/path
+    and a list of valid versions and languages,
+    extract and return any versions and languages and the remaining URL path
+    """
+
+    # Try parsing languages and versions
+    url_parts = request_path.split('/')[1:]
+    language = None
+    version = None
+
+    for part in url_parts:
+        if languages and part in languages:
+            language = part
+            url_parts.remove(part)
+        elif versions and part in versions:
+            version = part
+            url_parts.remove(part)
+
+    return (language, version, "/" + "/".join(url_parts))
+
+
+class YamlRegexMap:
+    def __init__(self, filepath):
+        """
+        Given the path to a YAML file of RegEx mappings like:
+
+            hello/(?P<person>.*)?: "/say-hello?name={person}"
+            google/(?P<search>.*)?: "https://google.com/?q={search}"
+
+        Return a list of compiled Regex matches and destination strings:
+
+            [
+                (<regex>, "/say-hello?name={person}"),
+                (<regex>, "https://google.com/?q={search}"),
+            ]
+        """
+
+        self.matches = []
+
+        if os.path.isfile(filepath):
+            with open(filepath) as redirects_file:
+                lines = yaml.load(
+                    redirects_file,
+                    Loader=yamlordereddictloader.Loader
+                )
+
+                if lines:
+                    for url_match, target_url in lines.items():
+                        # Make sure all matches start with slash
+                        if url_match[0] != '/':
+                            url_match = '/' + url_match
+
+                        self.matches.append(
+                            (re.compile(url_match), target_url)
+                        )
+
+    def get_target(self, url_path):
+        for (match, target) in self.matches:
+            result = match.fullmatch(url_path)
+
+            if result:
+                parts = {}
+                for name, value in result.groupdict().items():
+                    parts[name] = value or ''
+                return target.format(**parts)
```

Comment:
same here is possible to put the return out of the for loop?

Reply:
The same applies here. The redirects file can get thousands of lines long and regex matches are among some of the most expensive operations in Python, so I want to break the loop ASAP.